File: project-euler/145.py

def rev(n):
    r = 0
    while n > 0:
        r = r * 10 + (n % 10)
        n = n // 10
    return r


def even(n):
    return n % 2 == 0


def odd(n):
    return not even(n)


def solve(limit):
    memo = {}
    def reversible(i):
        if i == 0:
            return True
        elif i in memo:
            return memo[i]
        else:
            new_i = i // 10
            if new_i in memo:
                rec_res = memo[new_i]
            else:
                rec_res = reversible(new_i)
            result = odd(i % 10) and rec_res
            memo[i] = result
            return result

    count = 0
    for n in range(1, limit):
        if n % 1000000 == 0:
            print n
        rev_n = rev(n)
        if n % 10 == 0 or rev_n % 10 == 0:
            continue
        elif reversible(n + rev_n):
            count += 1
    return count


def main():
    print solve(1000000000)


main()

File: project-euler/145.hs

bound :: Int
bound = 1000000000

rev :: Int -> Int
rev = go 0
    where
        go r 0 = r
        go r n = let (d, m) = n `divMod` 10
                 in go (r * 10 + m) d

solve :: Int
solve = go bound 0
    where
        reversible :: Int -> Bool
        reversible n
            | n == 0            = True
            | n `mod` 10 == 0   = False
            | otherwise         = let (d, m) = n `divMod` 10
                                  in (odd m) && reversible d
        go :: Int -> Int -> Int
        go 0 c = c
        go n c = let r = rev n
                     incr = case (n `mod` 10 /= 0 && r `mod` 10 /= 0) of
                        True -> if reversible (r + n) then 1 else 0
                        False -> 0
                in go (n - 1) (c + incr)

main :: IO ()
main = print solve
